File: SCLAY1S/SCLAY1S.py
#================================================================
#
#                SCLAY1S MODEL FOR PYTHON3
#
#                     [BY Amardeep A]
#
#                  Created on June 12-2018
#================================================================


import numpy as np
import matplotlib.pyplot as plt

analysis = 1 # only values 0-Tri Dr, 1-Tri Undr, 2-Oedo

#===================== Input parameters =======================

sig0 = 20.0             # Horizontal (kPa)
sig1 = 30.0             # Vertical stress (kPa)
sig2 = 20.0             # Horizontal (kPa)
sig3 =  0.0             # Shear stress (kPa)
sig4 =  0.0             # Shear stress (kPa)
sig5 =  0.0             # Shear stress (kPa)

sig = np.array([[sig0],[sig1],[sig2],[sig3],[sig4],[sig5]])
strain = np.array([[0.0],[0.0],[0.0],[0.0],[0.0],[0.0]]) # rewinded to zero after initial loading to current stress state

kappa = 0.05            # Swelling index
lambda1 = 0.25          # Compression index
nu = 0.15               # poisson ratio
M = 1.5                 # critical state slope for compression
alpha0 = 0.58           # Initial inclination of yield surface
OCR = 0.0               # Overconsolidation ratio (OCR)
POP = 15.0              # Preoverburden pressure (kPa)
x0 = 3.0               # Initial bonding in the soil
e0 = 2.50               # Initial void ratio

# Derived parameters
phi = np.arcsin(3.0*M/(6.0 + M))
K0nc = 1.0 - np.sin(phi)
eta_k0 = 3.0*(1.0-K0nc)/(1.0+2.0*K0nc)
alphaK0 = (eta_k0**2+3.0*eta_k0-M**2)/3.0

print('Alpha: ',alpha0,' Alpha_K0nc: ',alphaK0)

#================== Calculate isotropic PCP ========================

sig_mod0 = (sig1+POP)*K0nc
sig_mod1 = (sig1+POP)
sig_mod2 = (sig1+POP)*K0nc
sig_mod3 = 0.0
sig_mod4 = 0.0
sig_mod5 = 0.0

#==================== Calculate stress invariants ===========================

pdash = (sig_mod0+sig_mod1+sig_mod2)/3.0

q = np.sqrt((3.0/2.0)*((sig_mod0-pdash)**2+(sig_mod1-pdash)**2+ \
    (sig_mod2-pdash)**2+2.0*(sig_mod3)**2+2.0*(sig_mod4)**2+2.0*(sig_mod5)**2))

pm = (((q-alpha0*pdash)**2)/(pdash*(M*M-alpha0*alpha0)))+pdash

pmi = pm/(1+x0)

#===================== Plotting the yield surface =============================

p_trial = np.linspace(0,pm,200)
p_triali = np.linspace(0,pmi,200)

q_trial1 = alpha0*p_trial + np.sqrt((pm-p_trial)*(M*M-alpha0*alpha0)*p_trial)
q_trial2 = alpha0*p_trial - np.sqrt((pm-p_trial)*(M*M-alpha0*alpha0)*p_trial)

q_triali1 = alpha0*p_triali + np.sqrt((pmi-p_triali)*(M*M-alpha0*alpha0)*p_triali)
q_triali2 = alpha0*p_triali - np.sqrt((pmi-p_triali)*(M*M-alpha0*alpha0)*p_triali)

pdash_given = (sig0+sig1+sig2)/3.0

q_given = np.sqrt((3.0/2.0)*((sig0-pdash)**2+(sig1-pdash)**2+ \
        (sig2-pdash)**2+2.0*(sig3)**2+2.0*(sig4)**2+2.0*(sig5)**2))

plt.figure(1)
plt.plot(p_trial,q_trial1,'-b')
plt.plot(p_trial,q_trial2,'-b')
plt.plot(p_triali,q_triali1,'-r')
plt.plot(p_triali,q_triali2,'-r')
plt.plot(pdash_given,q_given,'ro')
plt.plot([0,pdash+5],[0,(pdash+5)*M],'--m')
plt.plot([0,pdash+5],[0,-(pdash+5)*M],'--g')
plt.xlabel('Mean effective stress (kPa)')
plt.ylabel('Deviator stress (kPa)')
plt.title('Initial yield surface of SCLAY1S')
plt.grid(True)

#======================= Initialize the state variables =======================

StVar = np.array([[-(alpha0/3.0)+1.0],   # alpha_x
                 [(2.0*alpha0/3.0)+1.0], # alpha_y
                 [-(alpha0/3.0)+1.0],    # alpha_z
                 [0.0],                  # alpha_xy
                 [0.0],                  # alpha_yz
                 [0.0],                  # alpha_zx
                 [alpha0],               # alpha_scalar
                 [pm/(1+x0)],            # pmi
                 [x0],                   # x
                 [pm],                   # pm
                 [e0]])                  # current void ratio

# print(StVar)

#==============================================================================

strain_inc = 0.01/100           # strain increment
itr = 75                      # No of iterations

#==================== Initialization ==========================

# Updating at each iteration
p = np.zeros((itr,1))    # Mean effective stress (kPa)
p[0] = pdash_given

q = np.zeros((itr,1))    # Deviatoric stress (kPa)
u = np.zeros((itr,1))    # Pore pressure (kPa)
void = np.zeros((itr,1)) # Void ratio
epsV = np.zeros((itr,1)) # Volumetric strain
epsD = np.zeros((itr,1)) # Deviatoric strain


# Not updating for each iteration
D = np.zeros((6,6))      # Stiffness matrix
dfds = np.zeros((6,1))   # Yield surface derivative w.r.t stresses
dfdep = np.zeros((6,1))  # Yield surface derivative w.r.t volumetric strain



i = 0

# Calculate the initial yield surface

yield_s = ((q_given-alpha0*pdash_given)/(M*M-alpha0*alpha0))+(pdash_given-0.5*pm)**2-(0.5*pm)**2

while i<itr-1:
    K = nu*p[i]/kappa                           # Bulk modulus
    G = (3.0*K*(1.0-2.0*nu))/(2.0*(1.0+nu))    # Shear modulus
    if yield_s == 0:
        break
        # pc = (q[i]**2/ M**2 + p[i]**2) / p[i]
        # pm = (((q - alpha0 * pdash) ** 2) / (pdash * (M * M - alpha0 * alpha0))) + pdash
    else:
        pm = pm

    print('Iteration nr: ',i, '|| Yield value: ',yield_s,'|| pm: ',pm)

    # Considering isotropic elasticity inside the yield surface

    D[0, 0]=D[1, 1]=D[2, 2]= K+(4.0/3.0)*G
    D[0, 1]=D[0, 2]=D[1, 0]=D[1, 2]=D[2, 0]=D[2, 1]= K-(2.0/3.0)*G
    D[3, 3]=D[4, 4]=D[5, 5]= G


    if yield_s < 0:
        Dep = D
    else:
        break

    if analysis==0: # drained
        dstrain = np.array([[-Dep[1,0]/(Dep[1,1]+Dep[1,2])*strain_inc],
                   [strain_inc],
                   [-Dep[2,0]/(Dep[2,1]+Dep[2,2])*strain_inc],
                   [0.0],[0.0],[0.0]])

    elif analysis==1: # undrained
        dstrain = np.array([[-strain_inc*0.5],[strain_inc],[-strain_inc*0.5],[0.0],[0.0],[0.0]])
    elif analysis==2:
        dstrain = np.array([[0.0],[strain_inc],[0.0],[0.0],[0.0],[0.0]])

    dS = np.dot(Dep, dstrain)

    sig += dS

    strain += dstrain

    depsV = dstrain[0] + dstrain[1] + dstrain[2]
    depsD = (2.0 / 3.0) * (dstrain[0] - dstrain[1])

    i += 1

    p[i] = (sig[0] + sig[1] + sig[2]) / 3


    q[i] = np.sqrt((3.0/2.0)*((sig[0]-pdash)**2+(sig[1]-pdash)**2+ \
        (sig[2]-pdash)**2+2.0*(sig[3])**2+2.0*(sig[4])**2+2.0*(sig[5])**2))


    epsV[i] = epsV[i-1] + depsV
    epsD[i] = epsD[i-1] + depsD

    u[i] = ((sig0+sig1+sig2)/3.0)+(q[i]/3)-p[i]

    yield_s = ((q[i]-alpha0*p[i]/(M*M-alpha0*alpha0))+(p[i]-0.5*pm)**2-(0.5*pm)**2)

    print('yield_s: ', yield_s)

    if yield_s<0:
        ((q[i] - alpha0 * p[i] / (M * M - alpha0 * alpha0)) + (p[i] - 0.5 * pm) ** 2 - (0.5 * pm) ** 2)
    else:
        yield_s = 0




plt.figure(1)
plt.plot(p_trial,q_trial1,'-b')
plt.plot(p_trial,q_trial2,'-b')
plt.plot(p_triali,q_triali1,'-r')
plt.plot(p_triali,q_triali2,'-r')
plt.plot(pdash_given,q_given,'ro')
plt.plot(p,q,'-r')
plt.plot([0,pdash+5],[0,(pdash+5)*M],'--m')
plt.plot([0,pdash+5],[0,-(pdash+5)*M],'--g')
plt.xlabel('Mean effective stress (kPa)')
plt.ylabel('Deviator stress (kPa)')
plt.title('Initial yield surface of SCLAY1S')
plt.grid(True)
plt.show()

File: SCLAY1S/SCLAY1S-fortran_source/S-Clay1S.f90
! ----------------------------------------------------------------------------------------
! 'S-Clay1S': user-defined soil model for PLAXIS, Wiltafsky (2003)
!  modifications: to LAhey Fortran, Vogler (2006)
!  change subincrementing procedure for Plaxis 8.5, Vogler (2007)
!  Changed to automatic sup-stepping MNR algorithm by Siva (2010)
! ----------------------------------------------------------------------------------------

  Subroutine My_SCLAY	( IDTask, iMod, IsUndr, &
	                      iStep, iTer, iEl, Intp, &
		                  X, Y, Z, &
			              Time0, dTime, &
						  Props, Sig0, Swp0, StVar0, &
		                  dEps, D, BulkW, &
	                      Sig, Swp, StVar, ipl, &
		                  nStat, &
			              NonSym, iStrsDep, iTimeDep, iTang, &
				          iPrjDir, iPrjLen, &						
					      iAbort )
        
  IMPLICIT DOUBLE PRECISION (A-H,O-Z)
  common / SOAbort / iAbortSO
  common / iounit / iou
! ----------------------------------------------------------------------------------------
! User-defined soil model: S-Clay1S (iMod=1)
!
!  Depending on IDTask, 1 : Initialize state variables
!                       2 : calculate stresses
!                       3 : calculate material stiffness matrix
!                       4 : return number of state variables
!                       5 : inquire matrix properties
!                       6 : calculate elastic material stiffness matrix
!
!  Argument	I/O Type
!  -------- --- ----
!  IDTask   I   I    : see above
!  iMod     I   I    : model number (1..10)
!  IsUndr   I   I    : =1 for undrained, 0 otherwise
!  iStep    I   I    : Global step number
!  iter     I   I    : Global iteration number
!  iel      I   I    : Global element number
!  Intp     I   I    : Global integration point number
!  X        I   R    : X-Position of integration point
!  Y        I   R    : Y-Position of integration point
!  Z        I   R    : Z-Position of integration point
!  Time0    I   R    : Time at start of step
!  dTime    I   R    : Time increment
!  Props    I   R()  : List with model parameters
!  Sig0     I   R()  : Stresses at start of step
!  Swp0     I   R    : Excess pore pressure start of step
!  StVar0   I   R()  : State variable at start of step
!  dEps     I   R()  : Strain increment
!  D       I/O  R(,) : Material stiffness matrix
!  BulkW   I/O  R    : Bulkmodulus for water (undrained only)
!  Sig      O   R()  : Resulting stresses
!  Swp      O   R    : Resulting excess pore pressure
!  StVar    O   R()  : Resulting values state variables
!  ipl      O   I    : Plasticity indicator
!  nStat    O   I    : Number of state variables
!  NonSym   O   I    : Non-Symmetric D-matrix ?
!  iStrsDep O   I    : =1 for stress dependent D-matrix
!  iTimeDep O   I    : =1 for time dependent D-matrix
!  iPrjDir  I   I    : Project directory (ASCII numbers)
!  iPrjLen  I   I	 : Length of project directory name		
!  iAbort   O   I    : =1 to force stopping of calculation
! ----------------------------------------------------------------------------------------
  
  

  Integer			iPrjDir(iPrjLen)
  Double Precision	StVar(nStat)
  Double Precision	Sig(6),Sig_(6)
  Double Precision	Sig0(6),Sig0_(6)
  Double Precision	StVar0(nStat)
  Double Precision	dEps(6),dEps_(6)
  Double Precision	D(6,6)
  Double Precision	Props(50)

! ----------------------------------------------------------------------------------------
! Expected contents of Props(1..50)
! 
!  1 : kappa		slope of post yield compression line from e-ln_p'-diagram
!  2 : ny			Poisson's ratio
!  3 : lambda		slope of swelling line from e-ln_p'-diagram
!  4 : M			critical state M value (in triaxial compression)
!  5 : my			absolute effectiveness of rotational hardening
!  6 : beta			relative effectiveness of rotational hardening	
!  7 : a			absolute effectiveness of destructurational hardening
!  8 : b			relative effectiveness of destructurational hardening
!  9 : OCR			overconsolidation ratio
! 10 : POP			pre-overburden pressure like in PLAXIS
! 11 : e0			initial voil ratio
! 12 : alpha0		initial inclination of the yield surface
! 13 : x0			initial bonding effect	
! 14 : StepSize		for controlling the increment size	
! ----------------------------------------------------------------------------------------

  Double Precision	::	kappa	
  Double Precision	::	ny			
  Double Precision	::	lambda	
  Double Precision	::	M	
  Double Precision	::	my		
  Double Precision	::	beta		
  Double Precision	::	a		
  Double Precision	::	b			
  Double Precision	::	OCR
  Double Precision	::	POP
  Double Precision	::	e0	
  Double Precision	::	alpha0			
  Double Precision	::	x0	
  Double Precision  ::  StepSize			

  Double Precision	::	K0nc			! K0 value calculated from M
  Double Precision	::	phi				! phi calculated from M
  Double Precision	::	pm				! p'm = (1+x) * p'mi
  Double Precision	::	pmi				! p'mi = p'm / (1+x)
  Double Precision	::	Kdash			! compression modulus
  Double Precision	::	pdash			! mean effective stress
  Double Precision	::	q				! deviatoric stress
  Double Precision	::	alpha_scalar	! current alpha 
  Double Precision	::	alpha(6)		! current alpha 
  Double Precision	::	alpha_d(6)		! deviatoric fabric tensor
  Double Precision	::	sig_d(6)		! deviatoric stress vector
  Double Precision	::	term(6)			! auxilary vector for evaluation
  Double Precision	::  dGdSig(6)		! auxilary vector of derivatives
  Double Precision	::  dSigdLam(6)		! auxilary vector of derivatives
  Double Precision	::  dFdad(6)		! auxilary vector of derivatives
  Double Precision	::  daddEpsV(6)		! auxilary vector of derivatives
  Double Precision	::  daddEpsD(6)		! auxilary vector of derivatives
  Double Precision	::  dGdSigD(6)		! auxilary vector of derivatives
  Double Precision	::	xx				! current x 
  Double Precision	::	ee				! current void ratio
  Double Precision	::	alpha_K0		! alpha calculated from K0 and M
  Double Precision	::	eta_K0			! stress ratio for oedometric conditions

  Double Precision	::	dsig(6)			! stress increment
  Double Precision	::	Sig0_mod(6)		! modified initial stress state due to POP/OCR
  Double Precision	::	Sig_trial(6)	! trial stress state
  Double Precision	::	dSig_trial(6)	! trial stress increment

  Double Precision	::	dEps_trial(6)	! target strain increment 
  Double Precision	::	dEps_plastic(6)	! plastic strain increment 
  Double Precision	::	dEpsVol			! plastic volumetric strain increment
  Double Precision	::	dEpsDev(6)		! plastic deviatoric strain increment vector

  Double Precision	::	dpmi			! change of pmi
  Double Precision	::	dad(6)			! change of alpha_d
  Double Precision	::	dxx				! change of xx

  Integer			::	n_sub			! number of subincrements
  Double Precision	::	dEps_sub(6)		! strain increment for subincrementing
  Double Precision	::	dEps_norm		! norm of dEps for subincementing
  Double Precision	::	Sig0_sub(6)		! stress state at start of respective subincr.

  Logical			::	converged		! convergence for iteration
  Integer			::	ipl				! state of plasticity

  Character			::	fname*255		! file name for debugging
  Logical			::	IsOpen			! file status

	kappa		= Props(1)		
	ny			= Props(2)		
	lambda		= Props(3)
	M			= Props(4)	
	my			= Props(5)			
	beta		= Props(6)
	a			= Props(7)
	b			= Props(8)	
	OCR			= Props(9)		
	POP			= Props(10)
	e0			= Props(11)			
	alpha0		= Props(12)
	x0			= Props(13)


! ----------------------------------------------------------------------------------------
! Initialize state variables
! ----------------------------------------------------------------------------------------
    If (IDTask .Eq. 1) Then 

!      ! create file name for debugging
!	  fname=' '
!      Do i=1,iPrjLen
!        fname(i:i) = Char( iPrjDir(i) )
!      End Do
!	  fname= fname(:iPrjLen)//'\DATA.usrdbg.RR0'
!
!	  ! open debugging file
!	  Inquire(Unit= 1, Opened= IsOpen)
!	  If (.not.IsOpen) Then
!        Open(Unit= 1, File= fname, Position= 'append')
!        write(3,*)'starting next phase'							
!      End If
!      
!      Open( Unit= 2, File= fname(:iPrjLen)//'DATA.debug.RR0')
!       
!	  ! do IDTask1 only once
!	  If (StVar0(12)==123.) Return
!     
!	  ! create debugging file
!	  If (iEl==1.and.intp==1) Then
!	    Close(Unit= 1, Status= 'delete')
!        Open(Unit= 1,File= fname)
!        write(3,*)'initialization'								
!        Call WriVec(3,' Props...',Props,14)
!	  End If

!change made by jpgras: the debug files are put on a file called Running_files
!their unit number is changed        
      fname= 'Running_files\DATA.usrdbg.txt'        
        
      Inquire(Unit= 3, Opened= IsOpen)
   	  If (.not.IsOpen) Then
        Open(Unit= 3, File= fname, Position= 'append')
        Write(3,*)'starting next phase'							
      End If

      Open( Unit= 4, File = 'Running_files\DATA.debug.RR0')
   	  ! do IDTask1 only once
   	  If (StVar0(12)==123.) Return
      
   	  If (iEl==1.and.inte==1) Then
   	    Close(Unit= 3, Status= 'delete')
         Open(Unit= 3,File= fname)
         Write(3,*)'initialization'								
         Call WriVec(3,' Props...',Props,25)
       End If      
      
      ! checking input variables
	  if ((OCR.ne.0.).and.(POP.ne.0.)) then	! using POP and OCR together is not possible
        write(3,*)'ERROR: using POP and OCR together is not possible'								
	    Stop
	  end if
	  if (POP.lt.0.) then ! POP has to be positive (Compression positive in DLL)
        write(3,*)'ERROR: POP has to be negative (compression=positive in DLL)'								
	    Stop
	  end if
	  if (OCR.lt.0.) then ! negative OCR values are not possible 
        write(3,*)'ERROR: negative OCR values are not possible'								
	    Stop
	  end if
	  if (lambda==kappa) then ! dpmidEpsV not calculable
        write(3,*)'ERROR: dpmidEpsV not calculable - division by zero'								
	    Stop
	  end if

	  ! get K0nc value 
	  phi=asin(3.*M/(6.+M))
	  K0nc=1-sin(phi)
	  
	       
      Sig0(1) = Min(Sig0(1), -1.0d-8 )
	  Sig0(2) = Min(Sig0(2), -1.0d-8 )
      Sig0(3) = Min(Sig0(3), -1.0d-8 )

	  ! provide modified Sig0(1:6) in case of no POP and no OCR
	  Sig0_mod(1)= Sig0(2)*K0nc
	  Sig0_mod(2)= Sig0(2)
	  Sig0_mod(3)= Sig0(2)*K0nc
	  Sig0_mod(4)= 0.
	  Sig0_mod(5)= 0.
	  Sig0_mod(6)= 0.

	  ! write to file
	  If (iEl==1.and.intp==1) Then
        Call WriVec(3,' Sig0....',Sig0_mod,6)
	  End If

	  ! adjust Sig0(1:3) due to POP
	  if (POP.ne.0.) then
	  
        POP = -1*POP ! Change sign of POP to PLAXIS formulation of compression	  
	    
	    Sig0_mod(1)= (Sig0(2)+POP)*K0nc	
	    Sig0_mod(2)= (Sig0(2)+POP)	
	    Sig0_mod(3)= (Sig0(2)+POP)*K0nc	
	  end if

	  ! adjust Sig0(1:3) due to OCR
	  if (OCR.ne.0.) then
	    Sig0_mod(1)= (Sig0(2)*OCR)*K0nc
	    Sig0_mod(2)= (Sig0(2)*OCR)
	    Sig0_mod(3)= (Sig0(2)*OCR)*K0nc
	  end if

      ! change sign due to SCLAY formulation (compression positive)
	  Sig0_mod=(-1.)*Sig0_mod	  
	 
	  
	  ! write to file
	  If (iEl==1.and.intp==1) Then
        Call WriVec(3,' Sig0_mod',Sig0_mod,6)
        write(3,*)' K0nc POP OCR'								
        write(3,'(3(f8.3,x))') K0nc, POP, OCR								
	  End If

	  ! calculate stress invariants
	  pdash	= ( Sig0_mod(1) + Sig0_mod(2) + Sig0_mod(3) ) /3
	  q = sqrt(3./2.*( (Sig0_mod(1)-pdash)**2. + (Sig0_mod(2)-pdash)**2. &
					+ (Sig0_mod(3)-pdash)**2. + 2.*(Sig0_mod(4))**2. &
					+ 2.*(Sig0_mod(5))**2. + 2.*(Sig0_mod(6))**2. ))
	  
      ! checking input variables
	  if ((pdash==0.).or.(M**2.==alpha0**2.)) then	! pm not calculable
        write(3,*)'ERROR: pm not calculable - division by zero'								
	    Stop
	  end if

	  ! determine size of the initial yield curve
	  pm = ( (q-alpha0*pdash)**2. / (pdash*(M**2.-alpha0**2.)) ) + pdash	  
	  pm = Max(2.0d0, pm)
	 
	  !Change made by jpgras, umatwrapper is sending Stvar to incremental driver
      !then we replace StVar0 by stVar
	  ! pre-set state variables
	  StVar(1)	= -(alpha0/3.)+1.		! alpha_x
	  StVar(2)	= (2.*alpha0/3.)+1.		! alpha_y
	  StVar(3)	= -(alpha0/3.)+1.		! alpha_z
	  StVar(4)	= 0.					! alpha_xy
	  StVar(5)	= 0.					! alpha_yz
	  StVar(6)	= 0.					! alpha_zx
	  StVar(7)	= alpha0				! alpha_scalar
	  StVar(8)	= pm/(1.+x0)			! pmi
	  StVar(9)	= x0					! x
	  StVar(10)= pm					! pm
	  StVar(11)= e0					! e_current (current void ratio)
	  StVar(12)= 123.					! initialization is done
	  
	  ! check alpha0 towards alpha_K0
	  eta_K0=3*(1-K0nc)/(1+2*K0nc)
	  alpha_K0=(eta_K0**2+3*eta_K0-M**2)/3
	  
	  ! write to file
	  If (iEl==1.and.intp==1) Then
        Call WriVec(3,' StVar0',StVar0,12)
        write(3,*)' pdash q pm'								
        write(3,'(2x,3(f8.3,x))') pdash, q, pm								
        write(3,*)' alpha0 alpha_K0' 								
        write(3,'(2(f10.5,x))') alpha0, alpha_K0								
	  End If

	  ! output
	  if (iEl==1.and.intp==1) then
	    write (1,*) 'starting first step'	
        write (2,*) 'starting first step'
	  end if
		
    End If  !IDTask = 1
    

! ----------------------------------------------------------------------------------------
! calculate stresses
! ----------------------------------------------------------------------------------------
    If (IDTask .Eq. 2) Then 
   
	   Do i=1,6
          Sig0_(i) = -Sig0(i)
          dEps_(i) = -dEps(i)
          Sig_ (i) = -Sig(i)
        End Do
                
        CALL  S_CLAY1S_Sub	(      IsUndr, &
                                   Props, Sig0_, Swp0, StVar0, &
                                   dEps_, D, BulkW, &
                                   Sig_, Swp, StVar, ipl, &
                                   nStat )
        
        
        
        Do i=1,6
          Sig(i) = -Sig_(i)
        End Do
        
!		write (1,*) 'alpha', StVar(7)	
		

	End If  ! IDTask = 2

! ----------------------------------------------------------------------------------------
! calculate material stiffness matrix (D-matrix)
! ----------------------------------------------------------------------------------------
    If ( IDTask .Eq. 3 .Or. IDTask .Eq. 6 ) Then 
 
      ! get ny value
	  xNu = ny

	  ! determine shear modulus from Youngs' modulus
	  pdash = + ( Sig0(1) + Sig0(2) + Sig0(3) ) /3		! global mean stress
	  if ((pdash.gt.-1.).and.(pdash.le.0.)) pdash=-1.			  
	  if ((pdash.lt.1.).and.(pdash.gt.0.)) pdash=1.			  
	  ee = StVar0(11)									! get current void ratio
	  Kdash= -(1.+ee)*pdash/kappa						! ...compression modulus	
	  !Kdash= -(1.+e0)*pdash/kappa						! ...compression modulus
	  E= 3.*(1.-(2.*xNu))*Kdash							! E (Young's) from kappa

	  G   = 0.5*E/(1.+xNu)

	  ! compose liner elastic material stiffness matrix
      F1  = 2.*G*(1.-xNu)/(1.-2.*xNu)
      F2  = 2.*G*( xNu )/(1.-2.*xNu)
      Call MZeroR(D,36)
      Do i=1,3
        Do j=1,3
          D(i,j) = F2
        End Do
        D(i,i) = F1
        D(i+3,i+3) = G
      End Do

      !calculate bulk modulus
	  BulkW = 0
      If (IsUndr.Eq.1) Then
        xNu_U = 0.495d0
        Fac=(1+xNu_U)/(1-2*xNu_U) - (1+xNu)/(1-2*xNu)
        Fac=2D0*G/3D0  * Fac
        BulkW = Fac
      End If
    
	End If  ! IDTask = 3, 6

! ----------------------------------------------------------------------------------------
! get number of state parameters
! ----------------------------------------------------------------------------------------
    If (IDTask .Eq. 4) Then

	  nStat = 12	! (1)	alpha_x
					! (2)	alpha_y
					! (3)	alpha_z
					! (4)	alpha_xy
					! (5)	alpha_yz
					! (6)	alpha_zx
					! (7)	alpha_scalar
					! (8)	pmi
					! (9)	x
					! (10)	pm
					! (11)	e_current
					! (12)	initialization index

   End If  ! IDTask = 4

! ----------------------------------------------------------------------------------------
! get matrix attributes
! ----------------------------------------------------------------------------------------
    If (IDTask .Eq. 5) Then 

      NonSym   = 0  ! 1 for non-symmetric D-matrix		
      iStrsDep = 1  ! 1 for stress dependent D-matrix	
      iTang    = 0  ! 1 for tangent D-matrix
      iTimeDep = 0  ! 1 for time dependent D-matrix

    End If  ! IDTask = 5

  Return

  End Subroutine My_SCLAY